src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,81 @@ namespace cryptonote
       {
-        if( epee::file_io_utils::is_file_exist(path) )
+        boost::filesystem::directory_iterator end_itr;
+        for (boost::filesystem::directory_iterator iter(power_supply_class_path); iter != end_itr; ++iter)
         {
-          power_supply_path = path;
-          break;
+          boost::filesystem::path power_supply_path = iter->path();
+          if (boost::filesystem::is_directory(iter->path()))
+          {
+            std::string power_supply_present;
+            epee::file_io_utils::load_file_to_string(power_supply_path.string() + "/present", power_supply_present);
+            if (!boost::starts_with(power_supply_present, "1")) {
+              LOG_PRINT_L4("Power supply not present at " << power_supply_path);
+              continue;
+            }
+
+            std::string power_supply_type_path = power_supply_path.string() + "/type";
+            if (boost::filesystem::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type;
+              epee::file_io_utils::load_file_to_string(power_supply_type_path, power_supply_type);
+
+              // If there is an AC adapter that's present and online we can break early
+              if (boost::starts_with(power_supply_type, "Mains"))
+              {
+                std::string power_supply_online_path = power_supply_path.string() + "/online";
+                if (boost::filesystem::is_regular_file(power_supply_online_path))
+                {
+                  std::string power_supply_online;
+                  epee::file_io_utils::load_file_to_string(power_supply_online_path, power_supply_online);
+
+                  if (boost::starts_with(power_supply_online, "1"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else
+                  {
+                    continue;
+                  }
+                }
+              }
+              else if (boost::starts_with(power_supply_type, "Battery"))
+              {
+                std::string power_supply_status_path = power_supply_path.string() + "/status";
+                if (boost::filesystem::is_regular_file(power_supply_status_path))
+                {
+                  std::string power_supply_status;
+                  epee::file_io_utils::load_file_to_string(power_supply_status_path, power_supply_status);
+
+                  // Possible status are Charging, Full, Discharging, Not Charging, and Unknown
+                  // We are only need to handle negative states right now
+
+                  if (is_plugged == boost::logic::indeterminate &&
+                      (boost::starts_with(power_supply_status, "Charging") || boost::starts_with(power_supply_status, "Full")))
+                  {
+                    is_plugged = boost::logic::tribool(true);
+                  }
+                  else
+                  {
+                    is_plugged = boost::logic::tribool(false);
```

Comment:
Shouldn't this be indeterminate? E.g. if its not 'Charging' or 'Full' what other states can it be? From what I have read it can be discharging and unknown also at least.